### PR TITLE
New builtin types

### PR DESCRIPTION
**Pull request by lushoBarlett**
Short pr, because what I want to do next is related to parsing types, and it will be preparation for type variables actually.

### CODE REVIEW

File: src/typechecker.cpp [hunk 1]
```diff
@@ -7,20 +7,23 @@
...
+	m_core.new_type_function(0);  // 1  | int
+	m_core.new_type_function(0);  // 2  | float
+	m_core.new_type_function(0);  // 3  | string
+	m_core.new_type_function(1);  // 4  | array
+	m_core.new_type_function(1);  // 5  | dictionary
+	m_core.new_type_function(0);  // 6  | boolean
+	m_core.new_type_function(0);  // 7  | unit
+	m_core.new_type_function(-1); // 8  | sum
+	m_core.new_type_function(-1); // 9  | product
+	m_core.new_type_function(-1); // 10 | record
```
Comment (SebastianMestre) [inline src/typechecker.cpp:20]: This doesn't sound right to me.

A sum (or product, etc) type is a form of type function, it's not a particular type function.

I think `sum`, `product`, and `record` should be values in the `type_function_type` enum

Comment (SebastianMestre) [reply src/typechecker.cpp:20]: To sum up: I think a type function should either be a var, a builtin, a sum, a product, or a record

**Review by SebastianMestre — CHANGES_REQUESTED** [review]

**Comment by SebastianMestre** [comment]
Also, unrelated, just to leave a written record: we really need to change the name of all the `foo_bar_type` enums to `FooBarType` or, even better, `FooBarTag`

File: src/typechecker.cpp (hunk 1 above)
Comment (lushoBarlett) [reply src/typechecker.cpp:20]: Yea perhaps this is a better way to do it

**Review by SebastianMestre — APPROVED** [review]
Yes

**Comment by SebastianMestre** [comment]
makes sense

**Comment by lushoBarlett** [comment]
> Also, unrelated, just to leave a written record: we really need to change the name of all the foo_bar_type enums to FooBarType or, even better, FooBarTag

I could definitely do this next. I would rather call them Tags when appropriate. But I don't know where would they be exactly so you might have to step in.

**Pull request merged**